bazel: code coverage instrumentation for cockroach binary

RaduBerinde opened the pull request:
This change implements a special hook to dump coverage data (collected
through Bazel's infrastructure) from the Cockroach binary.

Sample cmdline to build in this mode:
```
bazel build --collect_code_coverage --bazel_code_coverage //pkg/cmd/cockroach:cockroach
```

The `--collect_code_coverage` flag enables Bazel-driven coverage
instrumentation. The `--bazel_code_coverage` flag is an alias for
`//build/toolchains:bazel_code_coverage_flag` which is conceptually
similar to setting the `bazel_code_cover` build tag (it causes the
build to use `code_cover_on.go`).

This build mode will allow us to collect coverage data from
roachtests: the nodes will run an instrumented binary and the
roachtest runner will pull the coverage data from the nodes.

Epic: none
Release mode: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/110280)
<!-- Reviewable:end -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110280)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rickystewart and @srosenberg)

___
*[`pkg/testutils/bazelcodecover/BUILD.bazel` line 14 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110280#-NdqOITUB41HG6UiG6kU:-NdqOITUB41HG6UiG6kV:b-p8gvbe) ([raw file](https://github.com/cockroachdb/cockroach/blob/8117496ddfed7524891f058cf54ffc29aa1766cb/pkg/testutils/bazelcodecover/BUILD.bazel#L14)):*
> ```Starlark
>     deps = select({
>         "//build/toolchains:bazel_code_coverage": [
>             "@io_bazel_rules_go//go/tools/bzltestutil:go_default_library",
> ```

I found it a little strange that I can use `tools/bzltestutil` when I'm actually importing`tools/coverdata` (but using`@io_bazel_rules_go//go/tools/coverdata:go_default_library` did not work)


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
bazel: code coverage instrumentation for cockroach binary

This change implements a special hook to dump coverage data (collected
through Bazel's infrastructure) from the Cockroach binary.

Sample cmdline to build in this mode:
```
bazel build --collect_code_coverage --bazel_code_coverage //pkg/cmd/cockroach:cockroach
```

The `--collect_code_coverage` flag enables Bazel-driven coverage
instrumentation. The `--bazel_code_coverage` flag is an alias for
`//build/toolchains:bazel_code_coverage_flag` which is conceptually
similar to setting the `bazel_code_cover` build tag (it causes the
build to use `code_cover_on.go`).

This build mode will allow us to collect coverage data from
roachtests: the nodes will run an instrumented binary and the
roachtest runner will pull the coverage data from the nodes.

Epic: none
Release mode: None

rickystewart commented on pkg/cmd/cockroach-short/main.go:
```diff
@@ -20,8 +20,10 @@ package main
 
 func main() {
+	bazelcodecover.MaybeInitCodeCoverage()
 	cli.Main()
```
1. Why not just put `MaybeInitCodeCoverage()` in `cli.Main()`?
2. Why not put it in `cockroach-oss` as well?

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110280)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rickystewart and @srosenberg)

___
*[`pkg/cmd/cockroach-short/main.go` line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110280#gh-1321971732:-Ne4hI-KCAzixv8qwLdo:b-302lfs) ([raw file](https://github.com/cockroachdb/cockroach/blob/70ddfa1d3bb822469ac8363612de5f524f307543/pkg/cmd/cockroach-short/main.go#L28)):*
<details><summary><i>Previously, rickystewart (Ricky Stewart) wrote…</i></summary><blockquote>

1. Why not just put `MaybeInitCodeCoverage()` in `cli.Main()`?
2. Why not put it in `cockroach-oss` as well?
</blockquote></details>

I saw something else (allocsim) using `cli.Main()`.

I figured we don't run roachtests against `cockroach-oss` but I can add it for consistency.


<!-- Sent from Reviewable.io -->

rickystewart reviewed (approved):
Great, thank you.

RaduBerinde commented:
TFTR!

RaduBerinde commented:
bors r+

srosenberg reviewed (commented):
Nice work!
